app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php

```diff
@@ -65,10 +91,24 @@ public function execute(\Magento\Framework\Event\Observer $observer)
             || $product->getIsChangedWebsites()
             || $product->dataHasChangedFor('visibility')
         ) {
-            if ($product->isVisibleInSiteVisibility()) {
+            if ($product->getVisibility() != Visibility::VISIBILITY_NOT_VISIBLE) {
                 $product->unsUrlPath();
                 $product->setUrlPath($this->productUrlPathGenerator->getUrlPath($product));
                 $this->urlPersist->replace($this->productUrlRewriteGenerator->generate($product));
+
+                //Remove any rewrite URLs for websites the product is not in
+                foreach ($this->storeManager->getWebsites() as $website) {
+                    $websiteId = $website->getWebsiteId();
+                    if (!in_array($websiteId, $product->getWebsiteIds())) {
+                        foreach ($this->storeWebsiteRelation->getStoreByWebsiteId($websiteId) as $storeId) {
+                            $this->urlPersist->deleteByData([
```

Comment:
Doing a lot of queries in a loop doesn’t look good enough solution. Can we use execute single query?